src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -21,47 +24,126 @@ internal class FileShareProvider: IArtifactProvider
+            {
+                FileShareActionRecord downloadRecord = clientTelemetry.CreateRecord<FileShareActionRecord>((level, uri, type) =>
+                    new FileShareActionRecord(level, uri, type, nameof(PublishArtifactAsync), context));
+
+                await clientTelemetry.MeasureActionAsync(
+                    record: downloadRecord,
+                    actionAsync: async () =>
+                    {
+                        return await PublishArtifactAsync(sourcePath, destPath, parallelCount, cancellationToken);
+                    }
+                );
+
+                // Send results to CustomerIntelligence
+                context.PublishTelemetry(area: PipelineArtifactConstants.AzurePipelinesAgent, feature: PipelineArtifactConstants.PipelineArtifact, record: downloadRecord);
             }
         }
 
-        public async Task PublishArtifactAsync(string sourcePath, string destPath, int parallelCount, CancellationToken cancellationToken)
+        public async Task<FileSharePublishResult> PublishArtifactAsync(string sourcePath, string destPath, int parallelCount, CancellationToken cancellationToken)
         {
-            await PublishArtifactUsingRobocopyAsync(this.context, sourcePath, destPath, parallelCount, cancellationToken);
+            return await PublishArtifactUsingRobocopyAsync(this.context, sourcePath, destPath, parallelCount, cancellationToken);
         }
 
-        private async Task CopyFileShareAsync(string downloadRootPath, string destPath, IEnumerable<string> minimatchPatterns, CancellationToken cancellationToken)
+        private async Task CopyFileShareAsync(
+            string downloadRootPath,
+            string destPath,
+            IEnumerable<string> minimatchPatterns,
+            CancellationToken cancellationToken,
+            FileShareDownloadResult result)
         {
             IEnumerable<Func<string, bool>> minimatcherFuncs = MinimatchHelper.GetMinimatchFuncs(minimatchPatterns, this.tracer);
-            await DownloadFileShareArtifactAsync(downloadRootPath, destPath, defaultParallelCount, cancellationToken, minimatcherFuncs);
+            await DownloadFileShareArtifactAsync(downloadRootPath, destPath, defaultParallelCount, cancellationToken, minimatcherFuncs, result);
         }
         
-        private async Task PublishArtifactUsingRobocopyAsync(AgentTaskPluginExecutionContext executionContext, string dropLocation, 
-            string downloadFolderPath, int parallelCount, CancellationToken cancellationToken)
+        private async Task<FileSharePublishResult> PublishArtifactUsingRobocopyAsync(
+            AgentTaskPluginExecutionContext executionContext,
+            string dropLocation,
+            string downloadFolderPath,
+            int parallelCount,
+            CancellationToken cancellationToken)
         {
+            Stopwatch watch = Stopwatch.StartNew();
             executionContext.Output(StringUtil.Loc("PublishingArtifactUsingRobocopy"));
```

Comment:
I just realized MeasureActionAsync in BlobStoreClentTelemetry already tracks the action with a timer so this is redundant